### PR TITLE
Update attributes in EEP 48

### PR DESCRIPTION
**Pull request by josevalim**
Currently, editors and documentation tooling must
parse metadata from several chunks in order to
function properly. The source is in the compilation chunk.
The behaviours a module implements is in the attributes
chunk. And the annotation of any definition requires
traversing abstract code.

With this commit, we introduce three new attributes,
`:behaviours`, `:source`, and `:source_anno`, so
this additional data can be standardized and easily located.

We also remove `edit_url` from the standard. As it is not
used by any of the official tooling, languages, or libraries.

### CODE REVIEW

**Comment by josevalim** [comment]
@garazdawi please let me know if you are happy with these changes and we can also provide a PR to OTP to attach this informtation to its Docs chunk.

**Comment by garazdawi** [comment]
Could you give an/some examples of how `source` and `source_anno` would look like? I'm especially wondering what paths in `source` and `source_anno -> file` are relative to?

File: lib/kernel/doc/guides/eep48_chapter.md [hunk 1]
```diff
@@ -145,8 +148,11 @@ information to each entry. This EEP documents the following metadata keys:
 - **`since := binary()`** - a binary representing the version such entry was
   added, such as `<<"1.3.0">>` or `<<"20.0">>`.
 
-- **`edit_url := binary()`** - a binary representing a URL to change the
```
Comment (lukaszsamson) [inline lib/kernel/doc/guides/eep48_chapter.md:148]: Is `edit_url` removal intentional?

File: lib/kernel/doc/guides/eep48_chapter.md [hunk 2]
```diff
@@ -145,8 +148,11 @@ information to each entry. This EEP documents the following metadata keys:
...
-  documentation itself.
+- **`source := binary()`** - the location of the source file for this module.
+  Applies only to the module metadata.
+
+- **`source_anno := erl_anno:anno()`** - annotation metadata of the source
```
Comment (lukaszsamson) [inline lib/kernel/doc/guides/eep48_chapter.md:154]: Range is usually more handy than position

File: lib/kernel/doc/guides/eep48_chapter.md (hunk 1 above)
Comment (josevalim) [reply lib/kernel/doc/guides/eep48_chapter.md:148]: Yes, per the description:

> We also remove edit_url from the standard. As it is not
> used by any of the official tooling, languages, or libraries.

File: lib/kernel/doc/guides/eep48_chapter.md (hunk 2 above)
Comment (josevalim) [reply lib/kernel/doc/guides/eep48_chapter.md:154]: When is a range used and when is a position used? Remember this is external information, not information about when a file is currently open in the editor (especially because, when a file is opened, its locations may have all changed already anyway).

**Comment by josevalim** [comment]
> Could you give an/some examples of how source and source_anno would look like? I'm especially wondering what paths in source and source_anno -> file are relative to?

@garazdawi it seems both Erlang and Elixir keep source as an absolute path, so that's how we should proceed?

```
iex(1)> :lists.module_info(:compile)
[
  version: ~c"8.4.3",
  options: [
    :debug_info,
    {:i, ~c"/buildroot/otp/lib/stdlib/src/../include"},
    {:i, ~c"/buildroot/otp/lib/stdlib/src/../../kernel/include"},
    :warn_missing_doc_function,
    :warn_missing_doc_callback,
    :warn_missing_spec_documented
  ],
  source: ~c"/buildroot/otp/lib/stdlib/src/lists.erl"
]
iex(2)> String.module_info(:compile)
[
  version: ~c"8.5",
  options: [
    :no_spawn_compiler_process,
    :from_core,
    :no_core_prepare,
    :no_auto_import,
    {:inline, [append_unless_empty: 2, prepend_unless_empty: 2]},
    {:inline, [duplicate: 2]},
    {:inline, [graphemes: 1]},
    {:inline, [next_grapheme: 1]},
    {:inline, [starts_with_string?: 3]},
    {:inline, [ends_with_string?: 3]},
    {:inline,
     [
       reverse_characters_to_binary: 1,
       grapheme_to_binary: 1,
       grapheme_byte_size: 1,
       codepoint_byte_size: 1
     ]}
  ],
  source: ~c"/Users/jose/OSS/elixir/lib/elixir/lib/string.ex"
]
```

The `source_anno` is going to be whatever is the function definition. The goal is to store the the second line in the example below:

```erlang
-doc "Example".
example() -> ok.
```

So a literal copy of whatever is in the AST. For the module definition, probably the line of `-module(foo)`.

File: lib/kernel/doc/guides/eep48_chapter.md (hunk 2 above)
Comment (lukaszsamson) [reply lib/kernel/doc/guides/eep48_chapter.md:154]: Almost everything in LSP operates on ranges (e.g. [go to definition](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.18/specification/#textDocument_definition) is `(documentURI, position) -> (documentURI, range)`. If you search for Postion in the spec it's ~140 results vs 430 for Range + 80 for Location (uri and range)

File: lib/kernel/doc/guides/eep48_chapter.md [hunk 3]
```diff
@@ -145,8 +148,11 @@ information to each entry. This EEP documents the following metadata keys:
 - **`since := binary()`** - a binary representing the version such entry was
   added, such as `<<"1.3.0">>` or `<<"20.0">>`.
 
-- **`edit_url := binary()`** - a binary representing a URL to change the
-  documentation itself.
+- **`source := binary()`** - the location of the source file for this module.
```
Comment (michalmuskala) [inline lib/kernel/doc/guides/eep48_chapter.md:151]: I think this should go a bit deeper into how this will interact with the `deterministic` option since capturing a local path here will introduce a new source of non-determinism of `.beam` files

**Comment by garazdawi** [comment]
Sounds good, feel free to open a PR adding these. Just make sure that `+deterministic` works.

File: lib/kernel/doc/guides/eep48_chapter.md [hunk 4]
```diff
@@ -145,8 +148,11 @@ information to each entry. This EEP documents the following metadata keys:
...
+  Applies only to the module metadata.
+
+- **`source_anno := erl_anno:anno()`** - annotation metadata of the source
+  code (opposite to the annotation of the documentation attribute).
```
Comment (michalmuskala) [inline lib/kernel/doc/guides/eep48_chapter.md:155]: I'm not sure what "annotation metadata of the source code" means. Can you give some example values?

File: lib/kernel/doc/guides/eep48_chapter.md (hunk 3 above)
Comment (michalmuskala) [reply lib/kernel/doc/guides/eep48_chapter.md:151]: I also believe, if we're storing location, we should likely call this `path` or similar. To me the name `source` somehow implies this will contain the entire source code as a string.

File: lib/kernel/doc/guides/eep48_chapter.md (hunk 4 above)
Comment (josevalim) [reply lib/kernel/doc/guides/eep48_chapter.md:155]: it is the Anno field (usually the second) in the Erlang Abstract Format. Do you have suggestions on how to describe it?

File: lib/kernel/doc/guides/eep48_chapter.md (hunk 2 above)
Comment (josevalim) [reply lib/kernel/doc/guides/eep48_chapter.md:154]: So in Elixir's case, it would go from `def ... end`? And in Erlang case, from the name until the dot? What happens when we only have a starting location, how are you filling it in? And, even if you had the range, would it actually be use by language servers? I never saw Go to Definition actually selecting the whole code. Which is also tricky if you have multiple clauses.

Comment (lukaszsamson) [reply lib/kernel/doc/guides/eep48_chapter.md:154]: It’s hard to answer. Either def to end or function name only. Maybe function name is more useful. Def to end use cases will most likely need parsing the source file anyway and we can get ranges from AST.
If the cursor is inside definition range editors do find references request so it definitely returns ranges e.g. in typescript VSCode.

Comment (josevalim) [reply lib/kernel/doc/guides/eep48_chapter.md:154]: > If the cursor is inside definition range editors do find references request so it definitely returns ranges e.g. in typescript VSCode.

Right, but in this case, you do have the file open anyway, so you can parse the range. Do we need the range for any operation when the file itself won't be open?

**Comment by garazdawi** [comment]
I wonder if it might also be a good idea to move the spec into the documentation metadata? So that all the docs can be rendered without looking into anything but the doc chunk?

**Comment by josevalim** [comment]
> I wonder if it might also be a good idea to move the spec into the documentation metadata? So that all the docs can be rendered without looking into anything but the doc chunk?

I thought about this, back when the EEP was written, and I could not decide, because if we are storing the whole spec as an AST, then we are effectively duplicating potentially large chunks of it. Could it be a problem? We could store only the string format but ExDoc will want to parse it to generate autolinks, and we would need to be able to parse both Elixir and Erlang versions. So I didn't see much benefit. Thoughts?

**Comment by garazdawi** [comment]
For `application/erlang+html` I stored the spec in the doc chunk because a single entry could refer to only part of a spec. So there it made sense. Also I just encountered that in a testcase that I would like to render the shell docs of a doc chunk of a different version than the currently running system, which is not possible right now as the spec is missing/incorrect.

When considering that doc chunks can also be external to the .beam file it makes even more sense as then you can build only the .chunk files and use those for rendering the complete docs.

For systems that are concerned about the file size, I think most strip doc chunks anyway? Or if they don't, then they can strip the debug info and still get complete documentation if the spec is included.

So in conclusion I think I would like to see the spec in there. Maybe the format should be as the AST with a function that can be called to render it into a string? For example

```
specification => {fun erl_pp:attribute/1, AST}
```

**Comment by josevalim** [comment]
Those are good points.

Storing the function is tricky because the function may not be available. For example, I may be looking at a Elixir module without fully loading Elixir itself (or `gleam` which does not have a runtime component after compilation). So I guess I just argued that, if we want to store specs, it should be textual?

**Comment by lukaszsamson** [comment]
> For application/erlang+html I stored the spec in the doc chunk

On elixir side that wasn't very useful as the spec was erlang code. A spec in typespec AST on the other hand can be easily translated back to elixir code

**Comment by garazdawi** [comment]
Any tool will most likely want want to use their own rendering functions for things anyway, for example in the shell we want to provide the current terminal width. So yes, a string will work just as well. We will have to rely on documentation for the tools to know how to parse the string.

File: lib/kernel/doc/guides/eep48_chapter.md (hunk 2 above)
Comment (garazdawi) [reply lib/kernel/doc/guides/eep48_chapter.md:154]: IMO we should stick with an `erl_anno` position as that is what is used in the `Anno` field of doc chunks.

Comment (josevalim) [reply lib/kernel/doc/guides/eep48_chapter.md:154]: @garazdawi I talked to @lukaszsamson and here are the concerns:

1. One function can actually have multiple clauses, with macros, attributes, etc. between them. So perhaps we should store the entry for each clause. So should we name the field `source_annos` and have it as a list?

2. In the future, storing the full range may be necessary. However, I think it should be trivial to add `end_location` to `erl_anno` in case we need it in the future. It already supports metadata attributes such as `text` and `record`, and a new field should be ok.

So my proposals are:

1. Update this PR to use `source_path` and `source_annos`
2. Send a PR adding `end_location` and `set_end_location` to `erl_anno`

WDYT?

Comment (josevalim) [reply lib/kernel/doc/guides/eep48_chapter.md:154]: For the `source_annos`, you can keep just one, if you don't want to keep them all, but we should support multiple if you want to.

Comment (josevalim) [reply lib/kernel/doc/guides/eep48_chapter.md:154]: Oh, `erl_anno` already has `end_location` but computes it from text. We could also allow it to just manually store it too. Fantastic.

Comment (garazdawi) [reply lib/kernel/doc/guides/eep48_chapter.md:154]: I don't mind having `source_annos`, though I'm not sure I see where that would be useful. Would ExDoc render multiple links to the source? Would ElixirLS use it to give multiple possible goto definition entries?

Regarding `end_location` to `erl_anno`, I would not mind such an addition, but that impacts the compiler folks a lot more than me so we will have to bring them into the discussion :) But feel free to open a PR and we can continue that discussion there.

Comment (josevalim) [reply lib/kernel/doc/guides/eep48_chapter.md:154]: The editor supports multiple entries, so I think we should consider that (given that's one of the main use cases) but in most cases we will have it as a single entry. I will submit a PR to add `erl_anno:end_location` and I will update this one based on the outcome of that. Be back soon. :)

**Comment by josevalim** [comment]
For those following this, PR for `end_location` is here: https://github.com/erlang/otp/pull/8966

**Comment by garazdawi** [comment]
Thanks!

**Comment by jonatanklosko** [comment]
FTR PR attaching the new information to the docs chunk: #8975.

**Comment by garazdawi** [comment]
I just realised that both Erlang and Elixir actually have paths as lists and not as a binary, so the doc chunks should do the same. #8975 already implements it as strings so the only thing to do is to update the eep-48 docs. I will post a PR for that soon.

**Comment by jonatanklosko** [comment]
@garazdawi great catch, thank you! In the Elixir implementation I actually did use a binary and I didn't realise the inconsistency. I've just a PR to Elixir to change to list (https://github.com/elixir-lang/elixir/pull/14071).

**Comment by josevalim** [comment]
Should we use `file:filename_all` or similar which can be either binary or char list?

**Comment by garazdawi** [comment]
I used file:filename to keep things simple. We can expand to _all later if needed.
